connector/protobuf/src/main/scala/org/apache/spark/sql/protobuf/utils/SchemaConverters.scala

```diff
@@ -92,14 +106,26 @@ object SchemaConverters {
             MapType(keyType, valueType, valueContainsNull = false).defaultConcreteType,
             nullable = false))
       case MESSAGE =>
-        if (existingRecordNames.contains(fd.getFullName)) {
+        // Setting the circularReferenceDepth to 0 allows the field to be recursed once, setting
+        // it to 1 allows it to be recursed twice, and setting it to 2 allows it to be recursed
+        // thrice. circularReferenceDepth value greater than 2 is not allowed. If the not
+        // specified, it will default to -1, which disables recursive fields.
+        val recordName = fd.getMessageType.getFullName
+        if (existingRecordNames.contains(recordName) &&
```

Comment:
Better to remove 'return' statement. 
How about"

    val recursiveDepth = existingRecordNames.getOrElse(recordName, 0)
    if (recursiveDepth == 0 ||  // No recursion
      (protobufOptions.circularReferenceDepth >= 0 
        && recursiveDepth <=  (protobufOptions.circularReferenceDepth + 1)) // recursion is within allowed limit.
         val newRecordNames = existingRecordNames + (recordName -> recursiveDepth + 1)) 
          ... 
    } else if (protobufOptions.circularReferenceDepth >= 0) { 
         // Recursion is allowed we reached limit. Truncate.
      return Some(StructField(fd.getName, NullType, ...)      
    } else { // Recursion is not allowed
       throw ...
    }

Reply:
Scratch the above suggestion. 
Instead you could add 'else' what you have and remove 'return'. That is simpler.

Reply:
@rangadi thanks for the review, I have made all changes you suggested.